### ISSUE TITLE
Avoid dnie_transmit_apdu in the dnie driver

### ISSUE THREAD

**Issue by rickyepoderi** [OpenSC/OpenSC#970]
Today I have started to look the code in order to delete the _dnie_transmit_apdu_ in _cwa-dnie.c_ as it was commented in pull #955. As it is known, inside the secure channel the dnie card transforms all the apdus in a CASE4 apdu. This way all the apdus through the SM need to call to the _sc_get_response_ method. The final response (after all the needed get_responses fill the data into the initial one) is the encoded apdu that should be decoded into the real apdu.

Looking to the code a typical _sc_transmit_apdu_ follows this path:

    --> sc_transmit_apdu(card, apdu): apdu.c
        |--> sc_transmit(card, pdu): apdu.c
             |--> sc_single_transmit(card, apdu): apdu.c
             |    |--> if SM_MODE_TRANSMIT sc_sm_single_transmit(card, apdu): sm.c
             |                             |--> get_sm_apdu(card, apdu, sm_apdu) (encode apdu)
             |                             |--> transmit(card, sm_apdu)
             |                             |--> free_sm_apdu(card, apdu, sm_apdu) (decode apdu)
             |--> if 0x61 sc_get_response(card, apdu, olen): apdu.c
                          |--> while data
                               |--> get_response(card, resp_len, resp) (implemented in iso7816.c)
                               |    |--> sc_transmit_apdu(card, apdu) (back to the beginning but with the new get_response apdu)
                               |--> accumulate the data to the first apdu

So the sm methods to get and free the encoded apdu are called in _sc_sm_single_transmit_. But in dnie we don't have the real encoded data until all the get_responses are transmitted and accumulated into the first apdu. That's why the _dnie_transmit_apdu_ exists, the _cwa_decode_response_ is called after all the get_responses are processed. It's impossible to emulate this behavior calling _free_sm_apdu_, just because it is never called when the initial apdu is fully filled by the get_response calls.

For DNIe we need another point of calling the _free_sm_apdu_ or another method that would be called after all the apdu is filled.

Please, correct me if I'm messing something.

**Comment by dengert** [comment]
I don't use SM, so I may be all wrong on this...

But it looks like free_sm_apdu does not decode anything, it just calls 
`card->sm_ctx.ops.free_sm_apdu` 

But any decoding of the response could be done after all the get_resposnes are done. It looks like that is what  cwa_decode_response should do. But only the cwa-dnie.c calls this routine. Could it have errors, or is being called in the wrong place?

For card-authenticate.c. free_sm_apdu is  authentic_sm_free_wrapped_apdu. It copies the first part of the received data from  sm_adpu to adpu.  The upper levels will do get_response to read the rest. 
But I don't see where it decodes all the response data or maybe it does not need to decrypt any?

For card-epass2003.c, free_sm_apdu is  epass2003_sm_free_wrapped_apdu.  It calls epass2003_sm_unwrap_apdu that calls decrypt_response but  decrypts only the date from the sm_apdu. And I don't see how it handles large response that need to do multiple get_response commands. 

So all three uses of SM do something different.

**Comment by rickyepoderi** [comment]
For what I understand the implementation of the SM in dnie is different to all the other cards. While dnie uses _cwa14890.h_  all the others uses _sm.h_. Inside _sm.h_ a _sm_card_operations_ struct is defined with several functions. Two of them are _get_sm_apdu_ (transforms the apdu before sending) and _free_sm_apdu_ (frees and converts apdu after sending). 

@frankmorgner has suggested several times that it would be nice to make dnie more compliant with the SM, a nice step is avoiding _dnie_transmit_apdu_ and using common _sc_transmit_apdu_. To achieve that we need that the dnie driver uses _get_sm_apdu_ and _free_sm_apdu_. But dnie needs to decode/free the apdu after all the get_responses are processed (when the encoded apdu is fully retrieved) and I see no possibility with the current opensc code (check the diagram of calls I put in the first paragraph).

**Comment by dengert** [comment]
Maybe what I am asking is the following in sm.c:167 correct:
`/* decode SM answer and free temporary SM related data */`
It leaves it up to the  card->sm_ctx.ops.free_sm_apdu to decode or not. 

Looking at your diagram,  you have the sc_sm_single_transmit calling transmit.  Does this "transmit" look for get_response and treat it any differently? 
Does this transmit  also include loop to call to get_response? 

 In ./card-dnie.c  dnie_sm_get_wrapped_apdu
`792         if (((plain->cla & 0x0C) == 0) && (plain->ins != 0xC0)) {`
it excludes get response. Is that correct?
Should these get_response be warped?

(I am marginally interested in SM and CWA14890. The NIST 800-73-4 defines Secure Messaging for PIV as optional. There are not cards available today, so I have not looked closely at it. But it looks a lot like  
CWA14890.)

**Comment by rickyepoderi** [comment]
The diagram is what I've understood looking to the code (maybe something is missing). 

>Looking at your diagram, you have the sc_sm_single_transmit calling transmit. Does this "transmit" look for get_response and treat it any differently?
>Does this transmit also include loop to call to get_response?

The _sc_single_transmit_ indirectly calls again to _sc_transmit_apdu_ recursively through the _sc_get_response_. The get response apdu just returns the data and therefore should never return a 61XX status. It doesn't treat it differently, simply that part of code is never followed in a get_response.

>In ./card-dnie.c dnie_sm_get_wrapped_apdu
>792 if (((plain->cla & 0x0C) == 0) && (plain->ins != 0xC0)) {
>it excludes get response. Is that correct?
>Should these get_response be warped?

Yes, it's correct. @germanblanco excluded the warping of get_responses apdus because that apdu should not be encoded. DNIe works like this (I don't know if it's the same for the rest of cards): it encodes any apdu into a CASE4 one; the encoded response is received through a 61XX (one or more get_responses are needed); and the obtained data is decoded into the real response apdu.

Maybe this behavior is specific for dnie and that's why it is not covered by general sm calls. IMHO dnie needs one more call or, better, another function to be called after all the get_responses are precessed and the initial apdu is complete.

**Comment by dengert** [comment]
Yes the code is very convoluted. Only thing I could suggest is to add `LOG_FUNC_CALLED`, `LOG_FUNC_RETURN` and maybe other LOG messages to see the code path better. 
And especially where a get_response is issued. 

I still question your diagram. Does the transmit(card, sm_apdu) include get_responses.
I think it does or it should do the get response.  And then the dnie_sm_free_wrapped_apdu could check if the sm_apdu completed or if returned 61XX.  

dnie_sm_free_wrapped_apdu  could append the data to the plain APDU. If the status is 61XX  then and return so get_response gets called in a loop.  dnie_sm_free_wrapped_apdu would get called again.  Then status 9000 (or error) is returned by the sm_apdu either by the initial command or a get_response,  the command is done and  dnie_sm_free_wrapped_apdu could then decode the data. This should work for short commands, or command requiring get_response. So no additional call is needed. 
 
I am just trying to give you some ideas. Hope this helps. 



in which case it could decode the

**Comment by frankmorgner** [comment]
@rickyepoderi, I think your analysis of the message flow in `apdu.c` is correct. However, quickly looking through the code, I was not able to verify your description of DNIe's command flow with the implementation. But if your description is correct, you could do the following:

1. In `dnie_sm_get_wrapped_apdu`, disable automatic sending of GET RESPONSE (`SC_APDU_FLAGS_NO_GET_RESP`; you should double check if this is needed).
2. rewrite `dnie_sm_free_wrapped_apdu` so that it checks for 61XX in response for a SM APDU. If found, disable SM (`SM_MODE_TRANSMIT`) and call `sc_get_response` to accumulate the response without applying SM. With the complete response, do decoding+decryption and return the plain text response data accordingly, reactivate SM. If 61XX is *not* the response for a SM APDU then directly do decoding+decryption.

The above rewrite should make it work with the existing SM framework in OpenSC. However, the whole topic of GET RESPONSE (and ENVELOPE) is very complicated in ISO7816-3/4. One simple thing I agree with you is that GET RESPONSE should be sent without SM even if `SM_MODE_TRANSMIT` is set. The German ID card does this as well. However, without a complete understanding about ISO7816 in this point, this could also just be card specific behavior... I would go with changing the code in the DNIe implementation described above.

**Comment by rickyepoderi** [comment]
Ok, let me think about it again. Option (1) may work but it's very specific, not very different than calling _dnie_transmit_apdu_. I have doubts about option (2), I need the first apdu completely filled and it's never passed like this to _free_sm_apdu_, no matter what I do, the call to that function is not in the proper point of time. But I'll double check it.

**Comment by dengert** [comment]
Do you have a opensc-debug.log to share? 
It would show where the get-responses are called and the 2 APDUs passed  to free_sm_apdu.

**Comment by dengert** [comment]
@viktorTarasov @frankmorgner 
cwa14890-1-2004 (The version I have) only has one line about get response
and no example: 

> "In the case the length of the public key cannot be coded in one byte and the card
does not support the extended length field concept according to ISO/IEC 7816-3
it is possible to use the GET RESPONSE command instead to retrieve the
remainder part of the public key."

The SM code appears to be used by only 3 drivers card-authenticate, card- epass2003 and dnie. 
is is possible that the the OpenSC SM code was written (or only tested with)  and extended APDU that 
do not require get response?

If so I can see why the dnie implementation is having so many issues. A single call to free_sm_apdu
should handle the recombining of the results of the main command and get response command(s).

**Comment by dengert** [comment]
Have a look at https://github.com/dengert/OpenSC/commit/b7138eb1d82e17859cd4649d9ad698caeba95f7e
it tries to have the get response done like any other APDU so that all the response data is in the sm_apdu before sm_ctx.ops.free_sm_apdu is called. 

This is untested and other drivers that use SM may need to be looked at.

**Comment by frankmorgner** [comment]
Indeed, I think all SM implementations don't use GET RESPONSE the way DNIe needs it. However, as depicted [above](https://github.com/OpenSC/OpenSC/issues/970#issuecomment-280881525), it only requires roughly three lines of codes to solve @rickyepoderi problem (please note that 1. and 2. are not standalone options; they both need to be done).

An even more elegant solution would be to implement the DNIe's card driver's `get_response` to temporarily disable SM. The upper layers should get the response (without SM) with no further modifications.

Currently, DNIe is creating its own SM stack, duplicating some of the code that's already present. I don't think you'll get a lot more functionality when refactoring, @rickyepoderi. Instead, I'm aiming for a smaller code base, simpler call stack and a more robust implementation if all drivers would be reusing the existing code base.

**Comment by dengert** [comment]
What I was trying to do, is implement what is **missing** in the base sm.c code.  apdu.c does a very nice job of handling the get response. Each SM caller should not have to implement it again.   

I would like to see get response handled in the base code.  There will be more cards that will require SM and will need get response handling. For example see: NIST 800-73-4,  Part 2,  Section 4.2.6 "Response with PIV Secure Messaging" Figure 7 " Chained Response under Secure Messaging"

The card-authenticate.c and card-epass2003.c need to be examined to see if they can use this code or if it interferes with what they do now.

**Comment by dengert** [comment]
@viktorTarasov  can you comment on how SM  should handle get_response?

**Comment by frankmorgner** [comment]
Since DNIe, the German ID card and the PIV card are implementing GET RESPONSE in SM in the same way, I implemented [my idea from above](https://github.com/OpenSC/OpenSC/issues/970#issuecomment-281012302). The [implementation](https://github.com/OpenSC/OpenSC/pull/972) is done in the ISO-7816 driver (since other cards may implement it differently) and gets on without introducing additional flags.

I'd be happy if you could look at the code and maybe test it with DNIe.

**Comment by dengert** [comment]
Is this a replacement for: https://github.com/dengert/OpenSC/commit/b7138eb1d82e17859cd4649d9ad698caeba95f7e ?

I agree that get_response is not be be sent via SM, as it is used to receive the results of the SM protected operation and you mod is needed. 

But it does **not** replace  https://github.com/dengert/OpenSC/commit/b7138eb1d82e17859cd4649d9ad698caeba95f7e 
which changes the diagram from above to:

```
--> sc_transmit_apdu(card, apdu): apdu.c
    |--> sc_transmit(card, pdu): apdu.c
         |--> sc_single_transmit(card, apdu): apdu.c
         |    |--> if SM_MODE_TRANSMIT sc_sm_single_transmit(card, apdu): sm.c
         |                             |--> get_sm_apdu(card, apdu, sm_apdu) (encode apdu)
         |*******************|--> sc_transmit_apdu(card, sm_apdu) *** recursive call to issue the sm_apdu as a normal APDU with get_response loop before call to free_sm_apdu in next line. free_sm_apdu will then have all the data from the original command and any get_responses. 
         |                             |--> free_sm_apdu(card, apdu, sm_apdu) (decode apdu)
Since the ********* call has done the get_response, the following should never get a 61XX
and a second get response will not be issued. 
         |--> if 0x61 sc_get_response(card, apdu, olen): apdu.c
                      |--> while data
                           |--> get_response(card, resp_len, resp) (implemented in iso7816.c)
                           |    |--> sc_transmit_apdu(card, apdu) (back to the beginning but with the new get_response apdu)
                           |--> accumulate the data to the first apdu
```

**Comment by rickyepoderi** [comment]
I think @dengert change may work. Please let me test it, but I have a very busy week, it's almost sure I won't have time to work on it until weekend.

@frankmorgner, I think your change wouldn't work. Although get response apdus don't follow the sm path, there is not call to _free_sm_apdu_ passing the original apdu and with the buffer filled. The only call to that function is done before the loop of get responses.

Thanks!

**Comment by frankmorgner** [comment]
Ah, now I can see why my approach is incomplete, thanks for the comments. And I also just realized that `sc_get_response` is not available outside of `apdu.c`. However, calling `sc_transmit_apdu` inside `sc_transmit_apdu` feels a bit itchy with the potential of endless recursion.

I think the real problem is the current separation of SM from the rest of the code, which makes cross cutting problems hard to deal with. A quick solution would be to introduce some code duplication (i.e. call `sc_get_response()` inside `sc_sm_single_transmit()` before `free_sm_apdu`. However, I recently decoupled `ENABLE_SM` from `ENABLE_OPENSSL`, which would allow a much tighter integration of SM into the core implementation. I'm not sure if someone finds the time to do this, but such refactoring would really be needed for future maintainability.

So in short, making the DNIe's SM implementation more elegant and more integrated, requires some changes to the core library. I'm not sure if this cause is worth the effort, but in the long run it could benefit the whole project.

**Comment by dengert** [comment]
Looking at the card_authenticate.c, card-epass2003.c and card-sc-hsm.c they all set SC_CARD_CAP_APDU_EXT. Thus there is no use of get_response by these modules. But card-dnie does not set SC_CARD_CAP_APDU_EXT, thus it is the first card driver to use SM with get_response.

**Comment by frankmorgner** [comment]
I just discovered that `sc_get_response()` has [some SM handling](https://github.com/OpenSC/OpenSC/blob/678f2bb1a65e5848dffc995f63e81d1f8092352f/src/libopensc/apdu.c#L468). Obviously, @viktorTarasov had a card that transmits GET RESPONSE APDU protected with SM which is the opposite of what the three cards mentioned above are doing. This complicates things a lot...

**Comment by rickyepoderi** [comment]
Remember that the real problem for dnie is when the call to _free_sm_apdu_ is done. Calling SM functions for the get response apdus can be solved (indeed @germanblanco already dealt with it just comparing if the apdu is a get_response  inside those functions, if it is a get response the apdu is not encoded). What dnie really needs is a call to _free_sm_apdu_ after the get responses. I see two solutions:

* Move the call to _free_sm_apdu_ outside _sc_sm_single_transmit_ and after the get responses (for example at the end of _sc_single_transmit_). You have to check if it affects other cards.
* Create another new function in the struct that just is just called if get responses are involved and after rearranging the data. This solution is global and does not affect other cards. But it is weirder,

**Comment by dengert** [comment]
Note that sc_sm_update_apdu_response is only called if get_response had an error. So this code may never have even been tested and never worked. 

We need @viktorTarasov  to comment on all of this. 

So I don't see anyway that a get response could be protected with SM. Get response is a special APDU. It is only used to retrieve data from a card after some other APDU was sent to the card and the card indicates only part of the result is being returned because of protocol limitations (T=0) or reader limitations, (no extended or other reader limitations).  The over all result is SM protected, but each get response is only used to retrieve a chunk of the results. 
 
The protected response could be returned in chunks with each chunk having its own MAC. But that is not what the PIV or DNIE appear to do. That would also make it more difficult for the card.  I would be surprised if there is a card that put a MAC on each chunk. 

Note since the card-authentic.c and card-epass2003 support SC_CARD_CAP_APDU_EXT there is 
most likely only one chunk returned and with T=1 it is returned by the first APDU without using get response.

**Comment by rickyepoderi** [comment]
Just as a proof of concept I have spent all the evening playing with @dengert solution. There are more issues because of the function _sc_set_le_and_transmit_ which is also called inside _sc_single_transmit_. But finally it worked with the dnie card.

For _sc_set_le_and_transmit_ I also needed to call to _sc_transmit_apdu_ and not an inner call but, more important, check [this change in the commit](https://github.com/rickyepoderi/OpenSC/commit/af2234db30121ed6b72726252dd40647ca533240#diff-6b74859ea2d4e1aad1102d7825bd537bL403). I think it is an error, cos your are avoiding _sc_get_response_ and not calling again to _sc_set_le_and_transmit_ (I think it's a bug that mixed both flags up).

You can see all my changes in the previous [commit](https://github.com/rickyepoderi/OpenSC/commit/af2234db30121ed6b72726252dd40647ca533240). I just made the _dnie_transmit_apdu_ to call directly to _sc_transmit_apdu_ (in the real change I would delete the function and replace all the calls), but for a PoC is more than enough.

With all those modifications dnie works perfectly and no _dnie_transmit_apdu_ would be needed. And please remember to change the wrong flag in _sc_set_le_and_transmit_, it's a bug for sure.

**Comment by dengert** [comment]
Great. Since I don't have any cards that do SM, please take  what I proposed and make whatever changes are needed and submit a pull request. 

@viktorTarasov  and any developers who  user SM should also test with their cards.

**Comment by rickyepoderi** [comment]
That [change](https://github.com/rickyepoderi/OpenSC/commit/af2234db30121ed6b72726252dd40647ca533240#diff-6b74859ea2d4e1aad1102d7825bd537bL414) is not needed. I did it first and then the one that changed the flag, with the second one is enough, but it is not a problem either.

If you are ok with this I will revert that call to the function _sc_single_transmit_ and just remove _dnie_transmit_apdu_ replacing it for _sc_transmit_apdu_. That will be the pull request. I think dnie can be improved a bit more later but I prefer to wait if the pull finally reaches master.

**Comment by dengert** [comment]
I am OK with whatever you want to do. Also look at https://github.com/OpenSC/OpenSC/pull/972 because it is related. 

What I was proposing, is you take https://github.com/dengert/OpenSC/commit/b7138eb1d82e17859cd4649d9ad698caeba95f7e
combine it with https://github.com/rickyepoderi/OpenSC/commit/af2234db30121ed6b72726252dd40647ca533240
rebase and squash into a commit for the base code and commit(s) for DNIE, and you submit the pull request.

**Comment by rickyepoderi** [comment]
Ok, it's done. Let's see what happens with the other SM cards. :smiley:

**Comment by viktorTarasov** [comment]
'It works for me'.
PR #972 tested with Gemalto IAS/ECC.

As it currently implemented, the SM for this card has no cases when more then one 'Get Response' per 'main' APDU  is needed.
APDU response is protected by SM.

**Comment by frankmorgner** [comment]
Thanks for your feedback! However, I still wonder what [this code](https://github.com/OpenSC/OpenSC/blob/678f2bb1a65e5848dffc995f63e81d1f8092352f/src/libopensc/apdu.c#L466-L469) was intended for:
```
		rv = card->ops->get_response(card, &resp_len, resp);
		if (rv < 0)   {
#ifdef ENABLE_SM
			if (resp_len)   {
				sc_log(ctx, "SM response data %s", sc_dump_hex(resp, resp_len));
				sc_sm_update_apdu_response(card, resp, resp_len, rv, apdu);
			}
#endif
			LOG_TEST_RET(ctx, rv, "GET RESPONSE error");
		}
```
Obviously it assumes that GET RESPONSE themselves are SM protected... Since it is no longer consistent with #975 and [maybe has never been triggered at all](https://github.com/OpenSC/OpenSC/issues/970#issuecomment-281345329), should we remove the code between the `ifdef`?
